snuba/datasets/transactions.py

```diff
@@ -0,0 +1,87 @@
+from snuba.clickhouse import (
+    ColumnSet,
+    DateTime,
+    IPv4,
+    IPv6,
+    LowCardinality,
+    Materialized,
+    Nested,
+    Nullable,
+    String,
+    Tuple,
+    UInt,
+    UUID,
+    WithDefault,
+)
+from snuba.datasets import DataSet
+from snuba.datasets.schema import ReplacingMergeTreeSchema
+from snuba.datasets.transactions_processor import TransactionsMessageProcessor
+
+
+class TransactionsDataSet(DataSet):
+    def __init__(self):
+        columns = ColumnSet([
+            ('project_id', UInt(64)),
+            ('event_id', UUID()),
+            ('trace_id', UUID()),
+            ('span_id', UInt(64)),
+            ('transaction_name', String()),
+            ('transaction_hash', Materialized(
+                UInt(64),
+                'cityHash64(transaction_name)',
+            )),
+            ('transaction_op', LowCardinality(String())),
+            ('start_ts', DateTime()),
+            ('start_ms', UInt(16)),
+            ('finish_ts', DateTime()),
+            ('finish_ms', UInt(16)),
+            ('duration', Materialized(
+                UInt(32),
+                '((finish_ts - start_ts) * 1000) + (finish_ms - start_ms)',
+            )),
+            ('platform', LowCardinality(String())),
+            ('environment', Nullable(String())),
+            ('release', Nullable(String())),
+            ('dist', Nullable(String())),
+            ('ip_tup', Tuple([
+                ('v4', Nullable(IPv4())),
+                ('v6', Nullable(IPv6())),
+            ])),
+            ('user', WithDefault(
+                String(),
+                "''",
+            )),
+            ('user_id', Nullable(String())),
+            ('user_name', Nullable(String())),
+            ('user_email', Nullable(String())),
+            ('tags', Nested([
+                ('key', String()),
+                ('value', String()),
+            ])),
+            ('contexts', Nested([
+                ('key', String()),
+                ('value', String()),
+            ])),
+            ('offset', UInt(64)),
+            ('partition', UInt(16)),
+            ('retention_days', UInt(16)),
+            ('deleted', UInt(8)),
+        ])
+
+        schema = ReplacingMergeTreeSchema(
+            columns=columns,
+            local_table_name='transactions_local',
+            dist_table_name='transactions_dist',
+            order_by='(project_id, toStartOfDay(start_ts), transaction_hash, start_ts, start_ms, span_id)',
```

Comment:
What were reasons behind choosing the ordering criteria?

Reply:
project_id and toStartOfDay(start_ts)  is to index by project and by time range
transaciton_hash helps grouping by transaciton type
span_id has to be part of the primary key to do the merge in case of deletion.

Reply:
Should we use the `span_id` or the `event_id` as the primary event identifier here?

Reply:
From a data fetch standpoint, when the product code tries to display one transaction in details, it works like events so it would use event_id and project_id.

From the conceptual standpoint, event_id is basically there only for backward compatibility.

In term of uniqueness this is how we generate span ids:
https://github.com/getsentry/sentry-python/blob/520158dab56a608eb4b678a2ca7006d8199eb9ce/sentry_sdk/tracing.py#L91

I am not sure this is guarantee to be unique per project but it should. Event_id certainly gives us the same guarantee it gives for errors. I can verify with @mitsuhiko  tomorrow

Reply:
after discussion with Armin, the identity is trace_ud + span_id.

@JTCunning  should we add a spearte index on event_id to facilitate the parts of the product that fetch individual transactions as sentry events ? (right now most of the product is based on the event table so they will fetch transacitons by project_id + event_id